src/type_url.rs

```diff
@@ -0,0 +1,19 @@
+//! Support for associating a type URL with a [`Message`].
+
+use crate::Message;
+
+/// Associate a type URL with the given [`Message]`.
+pub trait TypeUrl: Message {
+    /// Type URL for this [`Message`]. They take the form:
+    ///
+    /// ```text
+    /// /<package>.<TypeName>
+    /// ```
+    ///
+    /// For example:
+    ///
+    /// ```text
+    /// /foo.bar.baz.MyTypeName
+    /// ```
+    const TYPE_URL: &'static str;
```

Comment:
Just went with a `&'static str` for now, rather than trying to use a special type which validates the URL.

In my mind the goal is to get `prost-build` to spit these out eventually, in which case they should be well-formed. But maybe down the road it would be possible to add a URL type with a `const fn` parser.